polynote-kernel/src/main/scala/polynote/kernel/interpreter/python/PythonInterpreter.scala

```diff
@@ -245,20 +245,34 @@ class PythonInterpreter private[python] (
       |        return { 'globals': _globals, 'locals': _locals, 'types': types }
       |    except Exception as err:
       |        import traceback
-      |        typ, err_val, tb = sys.exc_info()
       |
-      |        # TODO: capture exception cause using __context__
-      |        trace = ArrayList()
-      |        for frame in traceback.extract_tb(tb):
-      |            trace.add(StackTraceElement(frame.filename.split("/")[-1], frame.name, frame.filename, frame.lineno))
-      |        result = { 'stack_trace': trace, 'message': getattr(err_val, 'message', str(err_val)), 'class': typ.__name__  }
+      |        def handle_exception(typ, err_val, tb):
+      |            trace = ArrayList()
+      |            for frame in traceback.extract_tb(tb):
+      |                trace.add(StackTraceElement(frame.filename.split("/")[-1], frame.name, frame.filename, frame.lineno))
+      |            result = {
+      |              'stack_trace': trace,
+      |              'message': getattr(err_val, 'message', str(err_val)),
+      |              'class': typ.__name__,
+      |              'err': err_val
+      |            }
       |
-      |        # TODO: it's a little ugly that we need to handle py4j stuff here :(
-      |        if typ.__name__ == 'Py4JJavaError':
-      |            result['py4j_error'] = err.java_exception._target_id
+      |            # Lifted from IPython.core.ultratb
+      |            def get_chained_exception(exception_value):
```

Comment:
Here's the source: https://github.com/ipython/ipython/blob/285e6a4400940ec6fcf5377be7141a1e42399e1b/IPython/core/ultratb.py#L533

I looked into whether we could just use `ultratb` from IPython but it wasn't straightforward.